src/cmake/modules/CUnit/src/main.c.in

```diff
@@ -9,23 +9,224 @@
+                opts.junit = true;
+                break;
+            case 'r':
+                if((i+1) < argc){
+                    opts.results = argv[++i];
+                    break;
+                }
+                /* FALLS THROUGH */
+            case 's':
+                if ((i+1) < argc) {
+                    opts.suite = argv[++i];
+                    break;
+                }
+                /* FALLS THROUGH */
+            case 't':
+                if ((i+1) < argc) {
+                    opts.test = argv[++i];
+                    break;
+                }
+                /* FALLS THROUGH */
+            default:
+                err = (CU_ErrorCode)256;
+                CU_set_error(err); /* Will print as "Undefined Errpr" */
+                break;
+        }
+    }
+
+    return err;
+}
+
+void
+add_suite(
+    const char *suite,
+    CU_InitializeFunc pInitFunc,
+    CU_CleanupFunc pCleanFunc)
+{
+    CU_pSuite pSuite;
+
+    pSuite = CU_get_suite(suite);
+    if (pSuite == NULL) {
+        pSuite = CU_add_suite(suite, pInitFunc, pCleanFunc);
+        CU_set_suite_active(pSuite, patmatch(opts.suite, suite));
+    }
+}
+
+#define ADD_SUITE(suite, init, clean) \
+  add_suite(#suite, init, clean)
+
+void
+add_test(
+    const char *suite,
+    const char *test,
+    CU_TestFunc pTestFunc,
+    bool enable)
+{
+    CU_pSuite pSuite;
+    CU_pTest pTest;
+
+    pSuite = CU_get_suite(suite);
+    pTest = CU_add_test(pSuite, test, pTestFunc);
+    CU_set_test_active(pTest, enable && patmatch(opts.test, test));
+}
+
+#define ADD_TEST(suite, test, enable) \
+  add_test(#suite, #test, CU_TestProxyName(suite, test), enable)
+
+/* CMake will expand the macro below to declare generated functions. */
+@decls@
+
+int main(int argc, char *argv[])
 {
     CU_ErrorCode err;
 
-    if ((err = cu_runner_init(argc, argv))) {
-        goto err_init;
+    if ((err = parse_options(argc, argv)) != CUE_SUCCESS) {
+        usage(argv[0]);
+        return err;
+    } else if (opts.print_help) {
+        help(argv[0]);
+        return CUE_SUCCESS;
+    }
+
+    if ((err = CU_initialize_registry()) != CUE_SUCCESS) {
+      fprintf(stderr, "CU_initialize_registry: %s\n", CU_get_error_msg());
+      return err;
+    }
+
+/* CMake will expand the macro below to register all suites and tests. */
+@defns@
+
+    CU_set_error_action(opts.error_action);
+
+    if (opts.automated) {
+        if (opts.results != NULL) {
+            CU_set_output_filename(opts.results);
+        }
+
+#if defined(HAVE_ENABLE_JUNIT_XML)
+        if (opts.junit) {
+            CU_automated_enable_junit_xml(CU_TRUE);
+        } else
+#endif
+        {
+            CU_list_tests_to_file();
+        }
+        CU_automated_run_tests();
+    } else {
+        CU_basic_set_mode(opts.mode);
+        CU_basic_run_tests();
+    }
+
+    err = CU_get_error();
+    if (err == CUE_SUCCESS) {
+        err = (CU_ErrorCode)(CU_get_number_of_failures() == 0 ? 0 : 256);
```

Comment:
Again: a ``CU_ErrorCode`` is not a program exit status: I would prefer the conversion to be done properly (CUE_SUCCESS and number-of-failures = 0, return 0; number-of-failures > 0, return 1; not CUE_SUCCESS, perhaps return 2 ... or whatever makes sense for the various values of ``CU_ErrorCode``).